### PR TITLE
Fix `config::Builder::set_credentials_provider` to override what's previsouly set

### PR DESCRIPTION
**Pull request by ysaito1001**
## Motivation and Context
Fixes https://github.com/awslabs/aws-sdk-rust/issues/973

## Description
Prior to the PR, if a customer explicitly passed a credentials provider to a client's config `Builder::set_credentials_provider`, what's passed did not override a credentials provider previously set ([actual use case](https://github.com/awslabs/aws-sdk-rust/issues/973#issuecomment-1827224049)).

While in general, we recommend customers single-source a credentials provider through [aws_config::ConfigLoader::credentials_provider](https://docs.rs/aws-config/1.0.1/aws_config/struct.ConfigLoader.html#method.credentials_provider), we should eliminate the said footgun in case they directly pass a credentials provider to a client's config `Builder`.

The PR reverts test signature updates in https://github.com/smithy-lang/smithy-rs/pull/3156 (in hindsight, having to update test signatures in that PR was a sign of regression).

## Testing
Added a Kotlin test to `CredentialProviderConfigTest.kt` to verify the fix

## Checklist

- [x] I have updated `CHANGELOG.next.toml` if I made changes to the AWS SDK, generated SDK code, or SDK runtime crates

----

_By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice._

### CODE REVIEW

File: aws/sdk/integration-tests/kms/tests/integration.rs [hunk 1]
```diff
@@ -52,7 +52,7 @@ async fn generate_random() {
             .header("content-type", "application/x-amz-json-1.1")
             .header("x-amz-target", "TrentService.GenerateRandom")
             .header("content-length", "20")
-            .header("authorization", "AWS4-HMAC-SHA256 Credential=ANOTREAL/20090213/us-east-1/kms/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-security-token;x-amz-target;x-amz-user-agent, Signature=703f72fe50c310e3ee1a7a106df947b980cb91bc8bad7a4a603b057096603aed")
+            .header("authorization", "AWS4-HMAC-SHA256 Credential=ANOTREAL/20090213/us-east-1/kms/aws4_request, SignedHeaders=content-length;content-type;host;x-amz-date;x-amz-target;x-amz-user-agent, Signature=53dcf70f6f852cb576185dcabef5aaa3d068704cf1b7ea7dc644efeaa46674d7")
```
Comment (jdisanti) [inline aws/sdk/integration-tests/kms/tests/integration.rs:55]: Why were these tests affected?

Comment (ysaito1001) [reply aws/sdk/integration-tests/kms/tests/integration.rs:55]: As described in the PR description, edited tests are now back to where they were before PR smithy-rs#3156; it was a mistake to update these signatures in that PR.

Comment (jdisanti) [reply aws/sdk/integration-tests/kms/tests/integration.rs:55]: I should have read more carefully 😅

**Review by MrCroxx — APPROVED** [review]
LGTM

File: aws/sdk/integration-tests/kms/tests/integration.rs (hunk 1 above)
Comment (rcoh) [reply aws/sdk/integration-tests/kms/tests/integration.rs:55]: yeah I think I thought something different was going on. I reviewed that PR too.

**Comment by ysaito1001** [comment]
Closing this on behalf of PR3278

**Pull request closed without merging**